app/WebParsing/ReqParser.hs

```diff
@@ -247,46 +288,107 @@ justParser = do
     markInfoParser = do
         Parsec.try (fmap Left percentParser <|> fmap Left letterParser <|> fmap Right infoParser)
 
--- parse for single course with or without cutoff OR a req within parentheses
-courseParser :: Parser Req
-courseParser = Parsec.between Parsec.spaces Parsec.spaces $ Parsec.choice $ map Parsec.try [
+-- Parses for single course or a group of programs with or without cutoff OR a req within parentheses
+-- Programs need to be parsed in groups because of the concatenation issue
+-- explained in the docstring of `programGroupParser`
+courseOrProgParser :: Parser Req
```

Comment:
Don't change `courseParser`, it also has a good logical function. Instead you can create a new parser that tries either `programOrParser` or `courseParser`. This will keep the different parsers easier to follow.